### PR TITLE
Add new [lfs_client].BATCH_SIZE and [server].LFS_MAX_BATCH_SIZE config settings.

### LINKED ISSUE #32306: LFS batch size should be configurable
### Feature Description

When configuring a mirror repository with LFS enabled, the size of the batch requests is 20. While troubleshooting batching, setting this to 1 would have been very helpful. While running this in a production mirror, having this batch size much higher could decrease mirroring times.

### Screenshots

_No response_

### PR DESCRIPTION
**Pull request by rremer**
This contains two backwards-compatible changes:
    * in the lfs http_client, the number of lfs oids requested per batch is loaded from lfs_client#BATCH_SIZE and defaulted to the previous value of 20
    * in the lfs server/service, the max number of lfs oids allowed in a batch api request is loaded from server#LFS_MAX_BATCH_SIZE and defaults to 'nil' which equates to the previous behavior of 'infinite'

This fixes #32306

### CODE REVIEW

**Comment by rremer** [comment]
fyi I ignored two failures under test-backend from `modules/zstd/zstd_test.go` as they are also failing on `main` and appear unrelated. My internal branch which is behind `main` by a few weeks has all tests passing.

Also, if there is a separate docs site I should be updating with the new gitea config, I would appreciate a pointer to where that lives.

**Comment by lunny** [comment]
You can send doc pull request here https://gitea.com/gitea/docs

**Comment by rremer** [comment]
related doc pr: https://gitea.com/gitea/docs/pulls/83

**Comment by lunny** [comment]
And maybe we should split lfs server configurations and lfs client configurations.

**Comment by rremer** [comment]
Fair enough @lunny , although it looks like the built-in lfs server doesn't have any constraint on batch size today: https://github.com/go-gitea/gitea/blob/main/services/lfs/server.go#L186

If you're in agreement, the way I'd go about this is:
1) change my proposed `LFS_BATCH_SIZE` to `LFS_CLIENT_BATCH_SIZE`
2) add a new `LFS_SERVER_MAX_BATCH_SIZE` config, and consume it early-on in the BatchRequest parsing of services/lfs/server.go, returning HTTP 413 Content Too Large if the slice length was greater than max batch size [git lfs reference doc at the bottom](https://github.com/git-lfs/git-lfs/blob/main/docs/api/batch.md)

If that sounds good, I'll update this branch with all that.

**Comment by lunny** [comment]
> Fair enough @lunny , although it looks like the built-in lfs server doesn't have any constraint on batch size today: [`main`/services/lfs/server.go#L186](https://github.com/go-gitea/gitea/blob/main/services/lfs/server.go#L186)
> 
> If you're in agreement, the way I'd go about this is:
> 
>     1. change my proposed `LFS_BATCH_SIZE` to `LFS_CLIENT_BATCH_SIZE`
> 
>     2. add a new `LFS_SERVER_MAX_BATCH_SIZE` config, and consume it early-on in the BatchRequest parsing of services/lfs/server.go, returning HTTP 413 Content Too Large if the slice length was greater than max batch size [git lfs reference doc at the bottom](https://github.com/git-lfs/git-lfs/blob/main/docs/api/batch.md)
> 
> 
> If that sounds good, I'll update this branch with all that.

I prefer to put the new configuration under `[lfs_client]`.

**Comment by rremer** [comment]
That's a bit confusing to me. All the other configs for lfs are under `[server]` today, but just to represent the options on the table here (please correct me if I misunderstood your suggestion):
```
[server]
LFS_START_SERVER = true
LFS_MAX_BATCH_SIZE = X

[lfs_client]
LFS_BATCH_SIZE = 20
```

And what I was proposing:
```
[server]
LFS_START_SERVER = true
LFS_CLIENT_BATCH_SIZE = 20
LFS_SERVER_MAX_BATCH_SIZE = X
```

We could make the argument for a config refactoring though, where we have `[server.lfs]` and `[server.lfs_client]`, but to maintain backwards-compatibility that would add some cruft to `modules/setting/lfs.go`.

**Comment by lunny** [comment]
> That's a bit confusing to me. All the other configs for lfs are under `[server]` today, but just to represent the options on the table here (please correct me if I misunderstood your suggestion):
> 
> ```
> [server]
> LFS_START_SERVER = true
> LFS_MAX_BATCH_SIZE = X
> 
> [lfs_client]
> LFS_BATCH_SIZE = 20
> ```
> 
> And what I was proposing:
> 
> ```
> [server]
> LFS_START_SERVER = true
> LFS_CLIENT_BATCH_SIZE = 20
> LFS_SERVER_MAX_BATCH_SIZE = X
> ```
> 
> We could make the argument for a config refactoring though, where we have `[server.lfs]` and `[server.lfs_client]`, but to maintain backwards-compatibility that would add some cruft to `modules/setting/lfs.go`.

Put all lfs configuration items under `[server]` is a legacy problem, I think it should be moved to `[lfs]` or `[server.lfs]` some day. So creating a new section for lfs client is better.

**Comment by rremer** [comment]
I thought that's what you were getting at, okay so what if we put *new* configs for lfs under a combined section, like:
```
[server]
LFS_START_SERVER = true

[lfs.server]
LFS_SERVER_MAX_BATCH_SIZE = X

[lfs.client]
LFS_CLIENT_BATCH_SIZE = 20
```

That would keep this pr simple without adding more config debt. In a separate PR I could extend the new lfs config class to handle old/new config sections and add warnings for their potential future deprecations.

**Comment by rremer** [comment]
Okay, I went with this (very similar to what we agreed upon above, but removing the duplicate LFS_ prefix):
```
[server]
LFS_START_SERVER = true

[lfs.server]
MAX_BATCH_SIZE = X

[lfs.client]
BATCH_SIZE = 20
```

I added a test for these defaults, and opened up #32332 to track the refactor/migration of the lfs configs under [server] to [lfs.server].

File: modules/setting/lfs.go [hunk 1]
```diff
@@ -22,11 +26,22 @@ var LFS = struct {
 	Storage *Storage
 }{}
 
+// LFSServer represents configuration for hosting Git LFS
+var LFSServer = struct {
+	MaxBatchSize int `ini:MAX_BATCH_SIZE`
```
Comment (lunny) [inline modules/setting/lfs.go]:
```suggestion
	MaxBatchSize int `ini:"MAX_BATCH_SIZE"`
```

Comment (rremer) [reply modules/setting/lfs.go]: I've been running `test-backend`... should probably start running `lint-backend` 😅 

Thanks for your patience as I become more familiar with this project.

File: services/lfs/server.go [hunk 2]
```diff
@@ -179,6 +179,12 @@ func BatchHandler(ctx *context.Context) {
 		return
 	}
 
+	batchSize := len(br.Objects)
```
Comment (lunny) [inline services/lfs/server.go]: LFSClient.BatchSize hasn't been used.

Comment (rremer) [reply services/lfs/server.go]: it's used the line below actually `batchSize > setting.LFSServer.MaxBatchSize`, but you're right that it's only used once. I was considering whether to put a debug line to print the batch size the client requested. Would you prefer:
* leave as-is
* add such a debug line
* remove this var and just get the length in-line for that same check

Comment (lunny) [reply services/lfs/server.go]: remove the lfs.client and the configuration items if no place use that.

Comment (rremer) [reply services/lfs/server.go]: okay, I think I probably don't want a log line, so I've moved this inline. We can add it back if we want logging later.

File: modules/lfs/http_client.go [hunk 3]
```diff
@@ -30,7 +29,7 @@ type HTTPClient struct {
 
 // BatchSize returns the preferred size of batchs to process
 func (c *HTTPClient) BatchSize() int {
-	return httpBatchSize
+	return setting.LFSClient.BatchSize
```
Comment (rremer) [inline modules/lfs/http_client.go:32]: I realized you were talking about this, thanks that was lost in the config refactor. Put it back.

**Comment by lunny** [comment]
The last one, please update `app.example.ini` file.

**Comment by rremer** [comment]
> The last one, please update `app.example.ini` file.

Done, although it was in .gitignore? Didn't see any doc on whether this is generated, so manually curated it and force-added it.

Related docs site pr: https://gitea.com/gitea/docs/pulls/83/files

**Review by lunny — APPROVED** [review]

**Comment by wxiaoguang** [comment]
The config system becomes more messy:


`[server].LFS_MAX_FILE_SIZE`
`[lfs].STORAGE_TYPE`
`[lfs.client].BATCH_SIZE`
`[lfs.server].MAX_BATCH_SIZE`

Would they cause problems for maintainers and site admin?

**Comment by rremer** [comment]
@wxiaoguang would you prefer:
```
[server].LFS_MAX_FILE_SIZE
[lfs].STORAGE_TYPE
[lfs].CLIENT_BATCH_SIZE
[lfs].SERVER_MAX_BATCH_SIZE
```

Or something else? I'll let you and @lunny hash it out, although I think we're all in agreement that we don't want more lfs-specific configs under `[server]`?

**Comment by wxiaoguang** [comment]
I have no clear idea at the moment and I don't mean against it. Just a question.

**Comment by lunny** [comment]
> @wxiaoguang would you prefer:
> 
> ```
> [server].LFS_MAX_FILE_SIZE
> [lfs].STORAGE_TYPE
> [lfs].CLIENT_BATCH_SIZE
> [lfs].SERVER_MAX_BATCH_SIZE
> ```
> 
> Or something else? I'll let you and @lunny hash it out, although I think we're all in agreement that we don't want more lfs-specific configs under `[server]`?

Maybe we `[lfs]` means `[lfs.server]` so we don't need `[lfs.server]`. Ideally, all configurations related to `lfs` server  should be under `[lfs]` and all configurations related to `lfs` client should be under `[lfs.client]`.

**Comment by wxiaoguang** [comment]
> Maybe we `[lfs]` means `[lfs.server]` so we don't need `[lfs.server]`. Ideally, all configurations related to `lfs` server should be under `[lfs]` and all configurations related to `lfs` client should be under `[lfs.client]`.

NO. `[lfs]` means `LFS Storage`, a mess. That's why I think new config options should be carefully designed. (again, just an opinion, no blocker)

**Comment by lunny** [comment]
> > Maybe we `[lfs]` means `[lfs.server]` so we don't need `[lfs.server]`. Ideally, all configurations related to `lfs` server should be under `[lfs]` and all configurations related to `lfs` client should be under `[lfs.client]`.
> 
> NO. `[lfs]` means `LFS Storage`, a mess. That's why I think new config options should be carefully designed. (again, just an opinion, no blocker)

But `[lfs]` can also have other options rather than storage configurations.

**Comment by rremer** [comment]
I've tried to summarize two of the suggested organizations of this config in #32332 just so we can visualize what they look like.

**Comment by wxiaoguang** [comment]
> > > 
> > 
> > 
> > NO. `[lfs]` means `LFS Storage`, a mess. That's why I think new config options should be carefully designed. (again, just an opinion, no blocker)
> 
> But `[lfs]` can also have other options rather than storage configurations.

No. Because the `lfs` section was never used like that, it was only used for "storage". Without a careful design there would be new regressions/bugs and more messy (there have been). The `lfs` section is a heavy technical debit in history, it should not be used for any other new purpose.

image

> I've tried to summarize two of the suggested organizations of this config in #32332 just so we can visualize what they look like.

Renaming config items involves a lot of backwards compatibility work ...... I am not sure if there would be some people spending time on it and make it right.

And keep in mind that it can't use `[lfs.client]` because the INI system has annoying "overriding" behavior. See the `TestConfigProviderBehaviors` test.


The only clear approach in my mind is `[server].LFS_MAX_BATCH_SIZE + [lfs_client].BATCH_SIZE`.

(that's just my opinion, maybe it doesn't need to hurry to do any change before there is a consensus)

**Comment by rremer** [comment]
Updated this branch and https://gitea.com/gitea/docs/pulls/83 with @wxiaoguang's preference of:
```
[server]
LFS_MAX_BATCH_SIZE

[lfs_client]
BATCH_SIZE
```

Should we then close/never #32332 ? Seems like we're committing to not refactoring the lfs configs and keeping it all under `[server]` going forward?

File: modules/setting/lfs.go [hunk 4]
```diff
@@ -10,23 +10,34 @@ import (
 	"code.gitea.io/gitea/modules/generate"
 )
 
-// LFS represents the configuration for Git LFS
+const (
+	LFSConfigSectionServer = "server"
+	LFSConfigSectionClient = "lfs_client"
+)
```
Comment (wxiaoguang) [inline modules/setting/lfs.go]: These don't look right. There was no such "config section const" in code, just use the names in code directly.

File: modules/setting/lfs.go [hunk 5]
```diff
@@ -10,23 +10,34 @@ import (
...
+// LFSClient represents configuration for mirroring upstream Git LFS
+var LFSClient = struct {
+	BatchSize int `ini:"BATCH_SIZE"`
+}{}
+
 func loadLFSFrom(rootCfg ConfigProvider) error {
-	sec := rootCfg.Section("server")
-	if err := sec.MapTo(&LFS); err != nil {
-		return fmt.Errorf("failed to map LFS settings: %v", err)
-	}
```
Comment (wxiaoguang) [inline modules/setting/lfs.go:29]: The old code should be kept since the config system is not refactored.

**Review by wxiaoguang — APPROVED** [review]

**Pull request merged**